### PR TITLE
chore: upgrade to TypeScript 3.5

### PR DESCRIPTION
**Pull request by raymondfeng**
Upgrade to TypeScript 3.5

I have to address a few breaking changes:

1. https://github.com/microsoft/TypeScript/issues/31661
2. Generic type is default to `unknown` instead of `{}` now.

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

File: examples/todo-list/src/__tests__/unit/controllers/todo-list-todo.controller.unit.ts [hunk 1]
```diff
@@ -98,7 +101,7 @@ describe('TodoController', () => {
 
   describe('patch()', () => {
     it('returns a number of todos updated', async () => {
-      patch.resolves([aChangedTodo].length);
+      patch.resolves({count: [aChangedTodo].length});
```
Comment (bajtos) [inline examples/todo-list/src/__tests__/unit/controllers/todo-list-todo.controller.unit.ts]: Nice! IIUC, the stricter type checks are detecting programming errors we were not aware of 👍

File: packages/boot/src/__tests__/unit/booters/datasource.booter.unit.ts [hunk 2]
```diff
@@ -3,14 +3,11 @@
...
-  RepositoryMixin,
-} from '@loopback/repository';
-import {expect, sinon, TestSandbox} from '@loopback/testlab';
-import {resolve} from 'path';
-import {DataSourceBooter, DataSourceDefaults} from '../../..';
+import { Application } from '@loopback/core';
+import { ApplicationWithRepositories, RepositoryMixin } from '@loopback/repository';
+import { expect, sinon, TestSandbox } from '@loopback/testlab';
+import { resolve } from 'path';
+import { DataSourceBooter, DataSourceDefaults } from '../../..';
```
Comment (bajtos) [inline packages/boot/src/__tests__/unit/booters/datasource.booter.unit.ts]: The spacing looks wrong to me, I think you forgot to reformat on save? (You can also run `npm run lint:fix`).

File: examples/todo/src/__tests__/unit/controllers/todo.controller.unit.ts [hunk 3]
```diff
@@ -100,7 +100,7 @@ describe('TodoController', () => {
 
     it('uses the provided filter', async () => {
       const find = todoRepo.stubs.find;
-      const filter: Filter = {where: {isCompleted: false}};
+      const filter: Filter<Todo> = {where: {isComplete: false}};
```
Comment (bajtos) [inline examples/todo/src/__tests__/unit/controllers/todo.controller.unit.ts]: Nice! 👍

**Comment by bajtos** [comment]
Fixed the problems in d92f66117

**Comment by bajtos** [comment]
Rebased on top of the latest `master`, squashed the commits into a single one and force-pushed into the feature branch. Let's wait for CI results before landing.

File: packages/boot/src/__tests__/unit/booters/datasource.booter.unit.ts (hunk 2 above)
Comment (raymondfeng) [reply packages/boot/src/__tests__/unit/booters/datasource.booter.unit.ts]: Hmm, there seems to be racing condition in VSCode. `Organize imports` adds spaces for import `{ ... }` while prettier should fix that.

**Pull request merged**